yield_plugin/src/yield_plugin.cpp

```diff
@@ -649,6 +696,32 @@ namespace yield_plugin
     return get_collision_time(original_tp, new_list, config_.intervehicle_collision_distance_in_m);
   }
 
+  std::map<uint32_t, rclcpp::Time> YieldPlugin::get_collision_times_concurrently(const std::vector<carma_perception_msgs::msg::ExternalObject>& external_objects,
+    const carma_planning_msgs::msg::TrajectoryPlan& original_tp)
+  {
+
+    std::vector<std::future<std::optional<rclcpp::Time>>> futures;
+    std::map<uint32_t, rclcpp::Time> collision_times;
+
+    // Launch asynchronous tasks to check for collision times
+    for (const auto& object : external_objects) {
+      futures.push_back(std::async(std::launch::async,[this, &original_tp, &object](){
+          return get_collision_time(original_tp, object);
+        })
+      );
+    }
+
+    // Collect results from futures and update collision_times
+    for (size_t i = 0; i < external_objects.size(); ++i) {
+      auto collision_time = futures[i].get(); // waits for the result
+      if (collision_time) {
+        collision_times[external_objects[i].id] = collision_time.value();
+      }
+    }
```

Comment:
We can use a range-based `for` loop if we refactor `futures`'s type to be a `std::map` (or `std::unordered_map`):

```cpp
std::unordered_map<uint32_t, std::future<std::optional<rclcpp::Time>>> futures;

// ...

for (const auto & external_object : external_objects) {
  if (const auto collision_time{futures.at(external_object.id)}) {
    collision_times[external_object.id] = collision_time.value();
  }
}

Reply:
Added this. I also fixed a minor build issue I missed.